Implement cli methods for `Payments` section of eclair API

mohitk108 opened the pull request:
The commands implemented are:

- [x] CreateInvoice
- [x] DeleteInvoice
- [x] ParseInvoice
- [x] PayInvoice
- [x] SendToNode
- [x] SendToRoute
- [x] GetSentInfo
- [x] GetRecievedInfo
- [x] ListReceivedPayments
- [x] GetInvoice
- [x] ListInvoice
- [x] ListPendingInvoices

Commit:
Implement the `createinvoice` command

Commit:
Implement the `deleteinvoice` command

Commit:
Implement the `parseinvoice` command

claddyk commented:
In the `createinvoice` command, the field `amountMsat` is given as optional. However if I do not pass it in the command, the field `amount` is missing from the `apiResponse`. For example, kindly refer this log:

```
claddy@claddy:~/Documents/eclair-cli$ eclair-cli.kexe createinvoice -p password --description IfeelSleepyy
error: api response could not be parsed: {"prefix":"lnbcrt","timestamp":1692055179,"nodeId":"03c5b161c16e9f8ef3f3bccfb74a6e9a3b423dd41fe2848174b7209f1c2ea25dad","serialized":"lnbcrt1pjd4d5tpp5mxm5rtarp9smkfcxar9w3qg8yc5ye5czw2e095cpnf0grmag8uesdq5f9nx2etv2dkx2ets09ussp5k2mulaqks4us7y3vmef2y7lr8gz55gh28jrfelgvdvx0qan7mwgqmqz9gxqrrsscqp79q7sqqqqqqqqqqqqqqqqqqqsqqqqqysgqw4mfpuckdnnhjj80kvqzph6f89dutuv0dvmv2fh5s7ccat3xjae9rmc22ugzzqz8ccm2dgw7q680alfl88t8c3l65ealmdxh39y3x8cptwvjp0","description":"IfeelSleepyy","paymentHash":"d9b741afa30961bb2706e8cae8810726284cd30272b2f2d3019a5e81efa83f33","paymentMetadata":"2a","expiry":3600,"minFinalCltvExpiry":30,"features":{"activated":{"trampoline_payment_prototype":"optional","payment_secret":"mandatory","basic_mpp":"optional","option_payment_metadata":"optional","var_onion_optin":"mandatory"},"unknown":[]},"routingInfo":[]} (code=1)
```
And when I pass the value of  `amountMsat` then it works fine,
```
claddy@claddy:~/Documents/eclair-cli$ eclair-cli.kexe createinvoice -p password --description IfeelSleepyy --amountMsat 999
{
  "type": "types.Invoice",
  "prefix": "lnbcrt",
  "timestamp": 1692055631,
  "nodeId": "03c5b161c16e9f8ef3f3bccfb74a6e9a3b423dd41fe2848174b7209f1c2ea25dad",
  "serialized": "lnbcrt9990p1pjd4wz0pp5j4vn0usn95n2rapus3ltxytr6sum5lrah2hq6yhphsxnaujz9m3qdq5f9nx2etv2dkx2ets09ussp58wudf374k2ntpyel7kmhnfyklnfrfep977ax75z9g50rnsfa2hpsmqz9gxqrrsscqp79q7sqqqqqqqqqqqqqqqqqqqsqqqqqysgq79wmpuhf20ha3z6f55sg0dje7k62w8vyn84zne5k9v8lm2vf6k2q792v7kgf4ch7rs2an3q33yks2gyfavdl7pejhqklr48mwar8l2cpnzek3g",
  "description": "IfeelSleepyy",
  "paymentHash": "955937f2132d26a1f43c847eb31163d439ba7c7dbaae0d12e1bc0d3ef2422ee2",
  "paymentMetadata": "2a",
  "expiry": 3600,
  "minFinalCltvExpiry": 30,
  "amount": 999,
  "features": {
    "activated": {
      "trampoline_payment_prototype": "optional",
      "payment_secret": "mandatory",
      "basic_mpp": "optional",
      "option_payment_metadata": "optional",
      "var_onion_optin": "mandatory"
    },
    "unknown": [
    ]
  },
  "routingInfo": [
  ]
}
```

Commit:
Implement the `payinvoice` command

Commit:
Implement the `sendtonode` command

Commit:
Implement the `sendtoroute` command

t-bast commented:
> In the createinvoice command, the field amountMsat is given as optional. However if I do not pass it in the command, the field amount is missing from the apiResponse.

That simply means you should make the `amountMsat` field optional in the response as well, doesn't it?

t-bast commented on src/nativeMain/kotlin/commands/PayInvoice.kt:
```diff
@@ -0,0 +1,73 @@
+            } else {
+                PayInvoiceResult(false, response)
+            }
+            Serialization.encode(result)
```
No, you should not rely on the response length, this is very brittle and will break if one of the field changes its encoding (or if there are optional tlvs that are filled).

You should instead first try to deserialize as a successful response, then if that fails try to deserialize as a failure response, then fallback to a generic error.

claddyk replied:
> No, you should not rely on the response length, this is very brittle and will break if one of the field changes its encoding (or if there are optional tlvs that are filled).
> 
> You should instead first try to deserialize as a successful response, then if that fails try to deserialize as a failure response, then fallback to a generic error.

Just a small doubt, 
Since what the API responds to us is a `uuid` which basically tells us `id` of an attempt to pay the invoice. Now there's a regex for `uuid` which is given on [stackoverflow](https://stackoverflow.com/questions/136505/searching-for-uuids-in-text-with-regex), that is `^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`

So we can just check if the response that we get is a `uuid` or not.
```
.map { response ->
                val result = if (response.matches(Regex("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"))) {
                    InvoiceResult(true, response)
                } else {
                    InvoiceResult(false, response)
                }
                Serialization.encode(result)
            }
```
The response after implementing this is:
```
claddy@claddy:~/Documents/eclair-cli$ eclair-cli.kexe payinvoice -p password --invoice lnbcrt5n1pjd6c09pp5c7ueze5hk8r2qupkcn7canx8dzv9qplhpye372hu4unjndmn3ftqdqufejhw5rp095kuan0d93k24r9wd6qsp5cxrvzp7l207he3lursph0nx2se9tdgryjyguz8c69ty2z7wute6qmqz9gxqrrsscqp79q7sqqqqqqqqqqqqqqqqqqqsqqqqqysgqwgzsh3nahdym4h0u7q85axffcarvem679x98uk3a43xgy8fwxdm85dy5ty0u46tthclf4082e7ph0jchz0z2ywhp6pf4dw9qrucdrcgpf6eqes --amountMsat 500
{
  "type": "types.InvoiceResult",
  "success": true,
  "message": "5687c787-191b-4184-be07-a32ccf1a015e"
}
```

claddyk commented:
> > In the createinvoice command, the field amountMsat is given as optional. However if I do not pass it in the command, the field amount is missing from the apiResponse.
> 
> That simply means you should make the `amountMsat` field optional in the response as well, doesn't it?

Yes, that surely works. I just wanted to check it once, because the `amountMsat` is given as optional in the parameters section. So, I thought that if we don't enter it, then also the response would follow the same format as given in documentation.

Commit:
Fix `payinvoice` execute method logic

- Make amount type nullable

Commit:
Implement the following commands

- `getsentinfo`
- `getreceivedinfo`
- `listreceivedpayments`
- `getinvoice`
- `listinvoices`
- `listpendinginvoices`

t-bast commented on src/nativeMain/kotlin/commands/GetSentInfo.kt:
```diff
@@ -0,0 +1,53 @@
+) {
+    private val paymentHash by option(
+        ArgType.String,
+        description = "The payment hash common to all payment attempts to be retrieved\tNo"
```
Wrong copy-paste?

t-bast commented on src/nativeMain/kotlin/commands/SendToNode.kt:
```diff
@@ -0,0 +1,66 @@
+            externalId,
+            pathFindingExperimentName,
+        ).map { response ->
+            val result = if (response.length == 36) {
```
You shouldn't rely on the response length here.

t-bast commented on src/nativeMain/kotlin/types/EclairApiTypes.kt:
```diff
@@ -28,4 +29,130 @@ data class NodeInfo(
+    val paymentMetadata: String,
+    val expiry: Int,
+    val minFinalCltvExpiry: Int,
+    val amount: Long? = null,
```
Do you really need to define a default value? It would be better if we didn't.

t-bast commented on src/nativeMain/kotlin/types/EclairApiTypes.kt:
```diff
@@ -28,4 +29,130 @@ data class NodeInfo(
+@Serializable
+data class ReceivePaymentStatus(
+    val type: String,
+    val amount: Long? = null,
```
Do you really need to define a default value? It would be better if we didn't.

claddyk replied:
> Do you really need to define a default value? It would be better if we didn't.

From what I have understood so far, when I have a nullable field in my data class `val amount: Long?` without a default value and the JSON that I'm trying to deserialize does not contain that field, the kotlinx-serializtion throws an exception because it expects a value but doesn't find it.
If I just provide a default value `val amount: Long? = null`, now the field is missing in the JSON, but still it could be deserialized now.
Also, if I do not provide a default value, then a few test cases fail because of the reason I mentioned. For example, here's a log which explains it.
```
> Task :cleanNativeTest
> Task :compileKotlinNative
> Task :nativeProcessResources NO-SOURCE
> Task :nativeMainKlibrary
> Task :compileTestKotlinNative
> Task :linkDebugTestNative
> Task :nativeTest FAILED

kotlin.AssertionError: Expected value to be null, but was: <ApiError(code=1, message=api response could not be parsed: [
  {
    "prefix": "lnbcrt",
    "timestamp": 1643719798,
    "nodeId": "028e2403fbfddb3d787843361f91adbda64c6f622921b19fb48f5766508bcadb29",
    "serialized": "lnbcrt1psljtrkpp5gqus3ys83p9cry4nj43ykjyvkuuhrcc5y45a6l569zwuc8pn2xxsdq0ydex2cmtd3jhxucsp543t76xycc9kpx4estwm6tjlpsht3m7d5jxe09tqnyjjux970y9lsmqz9gxqrrsscqp79qtzsqqqqqysgqxwh55ncvj3hv0cypm8vafku83gayzg7qa3zlu3lua76lk53t2m3rgt4d5qa04cfdd0f407p328c9el9xvy3r6z9um90m5pjaxrrazysqfkxfa7",
    "description": "#reckless",
    "paymentHash": "4039089207884b8192b395624b488cb73971e3142569dd7e9a289dcc1c33518d",
    "paymentMetadata": "2a",
    "expiry": 3600,
    "minFinalCltvExpiry": 30,
    "features": {
      "activated": {
        "payment_secret": "mandatory",
        "basic_mpp": "optional",
        "option_payment_metadata": "optional",
        "var_onion_optin": "mandatory"
      },
      "unknown": []
    },
    "routingInfo": []
  },
  {
    "prefix": "lnbcrt",
    "timestamp": 1643719828,
    "nodeId": "028e2403fbfddb3d787843361f91adbda64c6f622921b19fb48f5766508bcadb29",
    "serialized": "lnbcrt1psljty5pp5z2247c5w8cl30err7s9qx8rkejltq49mk8z6l5eqar7l43pehapsdq0ydex2cmtd3jhxucsp5hrcu5s0jftrmje4yavu580tlrq4mdmdevye8aevn6dsae4x5kejqmqz9gxqrrsscqp79qtzsqqqqqysgqwus3au5085tp02cwvpjexc5rq6qjezwxwr3yecdxr525qprv5zjxa98r69kx87cegavjw9u9299yfdhnes7mp4dztttyduchudvq64cq4pyx28",
    "description": "#reckless",
    "paymentHash": "12955f628e3e3f17e463f40a031c76ccbeb054bbb1c5afd320e8fdfac439bf43",
    "paymentMetadata": "2a",
    "expiry": 3600,
    "minFinalCltvExpiry": 30,
    "features": {
      "activated": {
        "payment_secret": "mandatory",
        "basic_mpp": "optional",
        "option_payment_metadata": "optional",
        "var_onion_optin": "mandatory"
      },
      "unknown": []
    },
    "routingInfo": []
  }
])>.
	at kotlin.Throwable#<init>(Unknown Source)
	at kotlin.Error#<init>(Unknown Source)
	at kotlin.AssertionError#<init>(Unknown Source)
	at kotlin.test.DefaultAsserter#fail(Unknown Source)
	at kotlin.test.Asserter#assertTrue(Unknown Source)
	at kotlin.test.Asserter#assertNull(Unknown Source)
	at kotlin.test#assertNull(Unknown Source)
	at kotlin.test#assertNull$default(Unknown Source)
	at commands.ListPendingInvoicesCommandTest#successful(Unknown Source)
	at commands.$ListPendingInvoicesCommandTest$test$0.$successful(Unknown Source)
	at commands.$ListPendingInvoicesCommandTest$test$0.$successful(Unknown Source)
	at kotlin.native.internal.test.BaseClassSuite.TestCase#run(Unknown Source)
	at kotlin.native.internal.test.TestRunner.run#internal(Unknown Source)
	at kotlin.native.internal.test.TestRunner.runIteration#internal(Unknown Source)
	at kotlin.native.internal.test.TestRunner#run(Unknown Source)
	at kotlin.native.internal.test#testLauncherEntryPoint(Unknown Source)
	at kotlin.native.internal.test#main(Unknown Source)
	at <global>.Konan_start(Unknown Source)
	at <global>.Init_and_run_start(Unknown Source)
	at <global>.0x0(Unknown Source)
	at <global>.__libc_start_main(Unknown Source)
	at <global>.0x0(Unknown Source)



commands.ListPendingInvoicesCommandTest.successful request FAILED
    kotlin.AssertionError at null:-1
1 test completed, 1 failed
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':nativeTest'.
> There were failing tests. See the report at: file:///home/claddy/Documents/eclair-cli/build/reports/tests/nativeTest/index.html
* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
* Get more help at https://help.gradle.org
BUILD FAILED in 9s
5 actionable tasks: 5 executed
```

t-bast replied:
Interesting, it's a pity that kotlinx-serialization isn't smarter, in that case we should keep the default value of `null`.

t-bast reviewed (commented):
Needs a rebase and a few comments to address, but otherwise looks good to me.

claddyk commented:
I've made the changes requested and ammend them into the last commit 7ef806613aae3cdbcccda713fd899e8e43fd7f78
